pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,264 @@
+		},
+	)
+}
+
+// NewKubeResourceQuota returns a new KubeResourceQuota admission plugin.
+func NewKubeResourceQuota(config *resourcequotaapi.Configuration) *KubeResourceQuota {
+	return &KubeResourceQuota{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+
+		userSuppliedConfiguration: config,
+
+		delegates: map[logicalcluster.Name]*resourcequota.QuotaAdmission{},
+	}
+}
+
+// KubeResourceQuota is an admission plugin that handles quota per logical cluster.
+type KubeResourceQuota struct {
+	*admission.Handler
+
+	// Injected/set via initializers
+	clusterWorkspaceLister tenancylisters.ClusterWorkspaceLister
+	kubeClusterClient      kubernetes.ClusterInterface
+	kubeInformers          informers.SharedInformerFactory
+	quotaConfiguration     quota.Configuration
+
+	// Manually set
+	userSuppliedConfiguration *resourcequotaapi.Configuration
+
+	lock      sync.RWMutex
+	delegates map[logicalcluster.Name]*resourcequota.QuotaAdmission
+}
+
+// ValidateInitialization validates all the expected fields are set.
+func (k *KubeResourceQuota) ValidateInitialization() error {
+	if k.clusterWorkspaceLister == nil {
+		return fmt.Errorf("missing clusterWorkspaceLister")
+	}
+	if k.kubeClusterClient == nil {
+		return fmt.Errorf("missing kubeClusterClient")
+	}
+	if k.kubeInformers == nil {
+		return fmt.Errorf("missing kubeInformers")
+	}
+	if k.quotaConfiguration == nil {
+		return fmt.Errorf("missing quotaConfiguration")
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &KubeResourceQuota{}
+var _ = initializers.WantsKcpInformers(&KubeResourceQuota{})
+var _ = initializers.WantsKubeInformers(&KubeResourceQuota{})
+var _ = initializers.WantsKubeClusterClient(&KubeResourceQuota{})
+var _ = initializer.WantsQuotaConfiguration(&KubeResourceQuota{})
+
+// Validate gets or creates a resourcequota.QuotaAdmission plugin for the logical cluster in the request and then
+// delegates validation to it.
+func (k *KubeResourceQuota) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	var delegate *resourcequota.QuotaAdmission
+
+	k.lock.RLock()
+	delegate = k.delegates[cluster.Name]
+	k.lock.RUnlock()
+
+	if delegate != nil {
+		return delegate.Validate(ctx, a, o)
+	}
+
+	delegate, err = k.getOrCreateDelegate(cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	return delegate.Validate(ctx, a, o)
+}
+
+// getOrCreateDelegate creates a resourcequota.QuotaAdmission plugin for clusterName.
+func (k *KubeResourceQuota) getOrCreateDelegate(clusterName logicalcluster.Name) (*resourcequota.QuotaAdmission, error) {
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	delegate := k.delegates[clusterName]
+	if delegate != nil {
+		return delegate, nil
+	}
+
+	// TODO cherry pick https://github.com/kubernetes/kubernetes/pull/110246 so we can get the right stop channel wired.
```

Comment:
any way to make this code break here when we pick or rebase?

Reply:
I don't see how the upstream PR gives you the right channel. Where should it come from? It must be connected to workspace deletion here?

Reply:
That's what I wrote about on the following line. What is the best way to do this, do you think? We could

- Use a new custom admission initializer that injects something that makes it easy to know when a cw is deleted
- Make a "mini controller" with a queue in this file that handles deletion events
- ... ?